### PR TITLE
Improvements in {get|set}_directive functions

### PR DESCRIPTION
**Pull request by martbab**
These should fix https://fedorahosted.org/freeipa/ticket/6460 and future-proof
the code a bit

### CODE REVIEW

File: ipapython/ipautil.py [hunk 1]
```diff
@@ -1294,15 +1294,17 @@ def unescape_seq(seq, *args):
 
 def escape_seq(seq, *args):
     """
-    escape (prepend '\\') all occurences of sequence in input strings
+    escape (prepend '\\') all occurences of sequence in input strings unless it
+    is already escaped.
```
Comment (pspacek) [inline ipapython/ipautil.py]: I think that the function called `escape_seq` should have no say in what is being escaped. Given that next commit modified `get_directive` to be inverse of `set_directive`, why is this change necessary at all?

File: ipaserver/install/installutils.py [hunk 2]
```diff
@@ -436,16 +436,27 @@ def format_directive(directive, value, separator, quotes, quote_char):
...
     A rather inefficient way to get a configuration directive.
+
+    :param filename: input filename
+    :param directive: directive name
+    :param separator: separator between directive and value
+    :param quote_char: the characters that are used in this particular config
+        file to quote values. This character will be stripped and unescaped
+        from the raw value.
+
+    :returns: The (unquoted) value if the directive was found, None otherwise
```
Comment (pspacek) [inline ipaserver/install/installutils.py]: I know this is not originally your code but KeyError exception souds more appropriate to me (for non-existing directives)...

File: ipaserver/install/installutils.py [hunk 3]
```diff
@@ -436,16 +436,27 @@ def format_directive(directive, value, separator, quotes, quote_char):
...
     """
     fd = open(filename, "r")
     for line in fd:
         if line.lstrip().startswith(directive):
             line = line.strip()
             result = line.split(separator, 1)[1]
-            result = result.strip('"')
+            result = result.strip(quote_char)
```
Comment (pspacek) [inline ipaserver/install/installutils.py]: I think that this usage of `strip` will cause problems if the text in config file is like `\"abc\"`. It will remove only the last `"` (which is incorrect anyway because `"` are part of the string and not quotation).

Comment (pspacek) [inline ipaserver/install/installutils.py]: BTW this usage of `strip` does not remove any whitespace (unless `quote_char` is a whitespace ...) so the commit message needs correction as well.

**Comment by pspacek** [comment]
Please see my in-line comments.

File: ipapython/ipautil.py (hunk 1 above)
Comment (martbab) [inline ipapython/ipautil.py]: Probably not, I was just thinking about future proofing the code. You are probably right that the function should be as dumb as possible.

File: ipaserver/install/installutils.py (hunk 2 above)
Comment (martbab) [inline ipaserver/install/installutils.py]: I would rather not alter the function's contract in this way, who know how many callers are relying on this particular behavior. If you open a ticket for this I can do it in separate PR.

File: ipaserver/install/installutils.py (hunk 3 above)
Comment (martbab) [inline ipaserver/install/installutils.py]: Regarding stripping the quote_char, I realized that the correct thing to do is actually to first unescape all quote chars and then do the strip as this is the inverse operation of set_directive. Do you agree with such a change?.

**Comment by martbab** [comment]
Please see my in-line replies to your in-line comments :).

File: ipaserver/install/installutils.py (hunk 2 above)
Comment (pspacek) [inline ipaserver/install/installutils.py]: No problem, I do not insist on this change.

File: ipaserver/install/installutils.py (hunk 3 above)
Comment (pspacek) [inline ipaserver/install/installutils.py]: Hmm, that depends. How is `quote_char` supposed to behave? Can it be used just once in each value like `    "this is my value"    `? Or is it possible to use it as in C strings like `    "first"  " part" "of " "string" ` producing string value `first part of string`?

Speaking of whitespace, the functions should be really inversible. Is there a reason to mock with the whitespace when reading values? (Just wondering out loud.)

Comment (pspacek) [inline ipaserver/install/installutils.py]: Thinking about it a bit more, I would argue that first use of `strip()` on line https://github.com/freeipa/freeipa/pull/347#discussion-diff-92979573R456 is incorrect anyway. The code first does `strip()` and then `split(separator, 1)[1]`.

What if result of `split(separator, 1)[1]` contains whitespace?

**Comment by martbab** [comment]
Can somebody take over the review of this PR please?

File: ipaserver/install/ipa_server_certinstall.py [hunk 4]
```diff
@@ -134,14 +134,18 @@ def install_http_cert(self):
         dirname = certs.NSS_DIR
 
         old_cert = installutils.get_directive(paths.HTTPD_NSS_CONF,
-                                              'NSSNickname')
+                                              'NSSNickname', quote_char="'")
```
Comment (tomaskrizek) [inline ipaserver/install/ipa_server_certinstall.py]: Why is it safe to assume the `NSSNickname` will always be enclosed in single quotes?

mod_nss allows values to be enclosed in double quotes. If double quotes were indeed used, they'd be interpreted as a part of the value, which is not intended. It would be the exact same issue as ticket 6460, except for double quotes instead of single quotes.

**Comment by tomaskrizek** [comment]
Please see my feedback in in-line comments.

File: ipaserver/install/ipa_server_certinstall.py (hunk 4 above)
Comment (martbab) [inline ipaserver/install/ipa_server_certinstall.py]: You are right. I think that I will delegate the quoting/unquoting transformations to separate functions. That will be more robust IMHO.

File: ipaserver/install/installutils.py [hunk 5]
```diff
@@ -380,33 +380,55 @@ def update_file(filename, orig, subst):
...
+
+def unquote_directive_value(value, quote_char):
+    """Unquote a directive value
+    :param value: string to unquote
+    :param quote_char: character to strip. All escaped occurences of
+        `quote_char` will be uncescaped during processing
+    :returns: processed value
+    """
+    unescaped_value = "".join(ipautil.unescape_seq(quote_char, value))
+    return unescaped_value.strip(quote_char)
```
Comment (tomaskrizek) [inline ipaserver/install/installutils.py]: Using `strip()` causes issues when the actual value starts or ends with the same type of quote as is used to escape it, e.g. `'cn=\'ABC\''` gets stripped to `cn='ABC`.

File: ipaserver/install/installutils.py [hunk 6]
```diff
@@ -380,33 +380,55 @@ def update_file(filename, orig, subst):
...
-   :param directive: directive name
-   :param value: value of the directive
-   :param quotes: whether to quote `value` in `quote_char`. If true, then
-        the `quote_char` are first escaped to avoid unparseable directives
-   :param quote_char: the character used for quoting `value`
+    :param filename: input filename
+    :param directive: directive name
+    :param value: value of the directive
+    :param quotes: whether to quote `value` in `quote_char`. If true, then
+        the `quote_char` are first escaped to avoid unparseable directives.
```
Comment (tomaskrizek) [inline ipaserver/install/installutils.py]: Please remove reference to quote_char from comment.

File: ipaserver/install/installutils.py [hunk 7]
```diff
@@ -416,33 +438,43 @@ def format_directive(directive, value, separator, quotes, quote_char):
...
 
+
 def get_directive(filename, directive, separator=' '):
     """
     A rather inefficient way to get a configuration directive.
+
+    :param filename: input filename
+    :param directive: directive name
+    :param separator: separator between directive and value
+    :param quote_char: the characters that are used in this particular config
```
Comment (tomaskrizek) [inline ipaserver/install/installutils.py]: Update doc here as well.

**Comment by tomaskrizek** [comment]
I still managed to find a an issue for certain edge cases. See inline comments for more info.

**Comment by martbab** [comment]
Thanks, I have fixed the docstrings. I have also made directive unquoting less silly.

**Comment by tomaskrizek** [comment]
I wasn't able to find any more issues with the quoting of certificate names. The directive quoting seems to work properly now.

**Comment by martbab** [comment]
Thank's, let's hope that all this code will be replaced by some proper configuration parsing mechanism in the future.

**Comment by martbab** [comment]
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/e1ed8b5eff40331ba532d37f3fb08814d8a55b77
https://fedorahosted.org/freeipa/changeset/517d43e78b8d8ea0b796a6ff6a379236eaae21df
https://fedorahosted.org/freeipa/changeset/2831b30e9a9de947481c058d8d32e174f951b1c0
https://fedorahosted.org/freeipa/changeset/86f4a93fb3aeb6742acab5abaa1c17b525ea4223
